docs/development_best_practices.md

```diff
@@ -0,0 +1,130 @@
+# Development Best Practices
+
+This document describes the best practices to follow when developing operators or other application components such as partitoners, stream codecs etc.
+
+## Operators
+
+These are general guidelines for all operators and the next sections talk about special considerations for input and output operators.
+
+* When writing a new operator to be used in an application, consider breaking it down into
+	* An abstract operator that encompasses the core functionality but leaves application specific schemas and logic to the implementation
+	* An optional concrete operator also in the library that extends the abstract operator and provides commonly used schema types such as strings, byte[] or POJOs.
+* Follow the conventions of the life cycle methods
+	* Do one time initialization of entities that apply for entire lifetime of the operator in the **setup** method, e.g., factory initializations. Initializations in **setup** are done in the container where the operator is deployed. Initializations in the constructor are discouraged as they lead to extra garbage in memory. This reason is as follows. The operator is instantiated on the client from where the application is being launched, then serialized and made available for launch. So the constructor is first called on the client and if it were to initialize any of the fields, that state would be saved during serialization. In the launch process the operator is deserialized. This would invoke the constructor again, which will initialize the fields but their state will get overwritten by the serialized state and the initial values would become garbage in memory.
+	* Do one time initialization for live entities in **activate** method, e.g., opening connections to a database server or starting a thread for asynchronous operations. The **activate** method is called right before processing starts so it is a better place to do these kind of initializations than at **setup** which can lead to a delay before processing data from the live entity.  
+	* Perform periodic tasks based on processing time in application windows
+	* Perform initializations needed for each application window in **beginWindow**
+	* Perform aggregations needed for each application window  in **endWindow**
```

Comment:
Seems like this only applies to some aggregations (like average) and not to others (such as sum). Clarifying that would avoid confusion.